### ISSUE TITLE
"TypeError: this.relative_url is not a function" occurs when "relative_link: true"

### ISSUE THREAD

**Issue by seaoak** [hexojs/hexo#2426]
Using the latest `master` branch of [hexojs/hexo](https://github.com/hexojs/hexo) repository:    
hexojs/hexo@5a7ea3e988dd6a0165792aef6ea3febad2897aff

When just only `relative_link: true` is modified from the default initial settings (`_config.yml`), `hexo generate` command shows an error message and *normally finishes*.

The procedure for reproducing this issue is as follows:

```bash
$ hexo init blog
$ cd blog
$ npm install
$ npm install deep-assign
$ npm ls --depth 0
hexo-site@0.0.0 /home/foobar/blog
tqq deep-assign@2.0.0 extraneous
tqq hexo@3.2.2
tqq hexo-generator-archive@0.1.4
tqq hexo-generator-category@0.1.3
tqq hexo-generator-index@0.2.0
tqq hexo-generator-tag@0.2.0
tqq hexo-renderer-ejs@0.2.0
tqq hexo-renderer-marked@0.2.11
tqq hexo-renderer-stylus@0.3.1
mqq hexo-server@0.2.0

npm ERR! extraneous: deep-assign@2.0.0 /home/foobar/blog/node_modules/deep-assign
$ cd node_modules
$ rm -rf hexo
$ git clone ssh://git@github.com:22/hexojs/hexo.git
$ cd hexo
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
$ git log -1 --oneline | cat
5a7ea3e Replace new lines by spaces in description (#2354)
$ cd ../..
$ hexo version
hexo: 3.2.2
hexo-cli: 1.0.2
os: Linux 4.4.0-64-generic linux x64
http_parser: 2.7.0
node: 7.6.0
v8: 5.5.372.40
uv: 1.11.0
zlib: 1.2.11
ares: 1.10.1-DEV
modules: 51
openssl: 1.0.2k
icu: 58.2
unicode: 9.0
cldr: 30.0.3
tz: 2016j
$ grep '"version":' node_modules/hexo*/package.json
node_modules/hexo-front-matter/package.json:  "version": "0.2.3"
node_modules/hexo-fs/package.json:  "version": "0.1.6"
node_modules/hexo-generator-archive/package.json:  "version": "0.1.4"
node_modules/hexo-generator-category/package.json:  "version": "0.1.3"
node_modules/hexo-generator-index/package.json:  "version": "0.2.0"
node_modules/hexo-generator-tag/package.json:  "version": "0.2.0"
node_modules/hexo-i18n/package.json:  "version": "0.2.1"
node_modules/hexo-log/package.json:  "version": "0.1.2"
node_modules/hexo/package.json:  "version": "3.2.2",
node_modules/hexo-pagination/package.json:  "version": "0.0.2"
node_modules/hexo-renderer-ejs/package.json:  "version": "0.2.0"
node_modules/hexo-renderer-marked/package.json:  "version": "0.2.11"
node_modules/hexo-renderer-stylus/package.json:  "version": "0.3.1"
node_modules/hexo-server/package.json:  "version": "0.2.0"
node_modules/hexo-util/package.json:  "version": "0.6.0"
$
```

At the default initial settings, `hexo generate` command successfully finishes:

```bash
$ (hexo clean --debug && hexo generate --debug) >z 2>&1
$ echo $?
0
$ ls public/
2017/  archives/  css/  fancybox/  index.html  js/
$
```

When just only `relative_link: true` is modified in `_config.yml`, `hexo generate` command finishes *normally* but `index.html` is not generated:

```bash
$ cp -a _config.yml _config.yml.org
$ vi _config.yml
$ diff _config.yml.org _config.yml
38c38
< relative_link: false
---
> relative_link: true
$ (hexo clean --debug && hexo generate --debug) >z 2>&1
$ echo $?
0
$ ls public/
css/  fancybox/  js/
$
```

The log file "z" contains an error message as follows:

```
16:08:22.000 ERROR Process failed: _posts/hello-world.md
TypeError: this.relative_url is not a function
    at Hexo.urlForHelper (/home/foobar/blog/node_modules/hexo/lib/plugins/helper/url_for.js:28:17)
    at Model._Document.Document.<anonymous> (/home/foobar/blog/node_modules/hexo/lib/models/post.js:60:31)
    at Model._Document.Document.permalink (/home/foobar/blog/node_modules/warehouse/lib/types/virtual.js:70:22)
    at Model._Document.Document.Document.toObject (/home/foobar/blog/node_modules/warehouse/lib/document.js:79:42)
    at Database._Model.Model.Model._insertOne (/home/foobar/blog/node_modules/warehouse/lib/model.js:189:21)
    at /home/foobar/blog/node_modules/warehouse/lib/model.js:214:17
    at tryCatcher (/home/foobar/blog/node_modules/bluebird/js/release/util.js:16:23)
    at /home/foobar/blog/node_modules/bluebird/js/release/using.js:185:26
    at tryCatcher (/home/foobar/blog/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:510:31)
    at Promise._settlePromise (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:567:18)
    at Promise._settlePromise0 (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:691:18)
    at Promise._fulfill (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:636:18)
    at PromiseArray._resolve (/home/foobar/blog/node_modules/bluebird/js/release/promise_array.js:125:19)
    at PromiseArray._promiseFulfilled (/home/foobar/blog/node_modules/bluebird/js/release/promise_array.js:143:14)
    at Promise._settlePromise (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:572:26)
    at Promise._settlePromise0 (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/home/foobar/blog/node_modules/bluebird/js/release/promise.js:691:18)
    at Async._drainQueue (/home/foobar/blog/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/home/foobar/blog/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/home/foobar/blog/node_modules/bluebird/js/release/async.js:17:14)
```

I read the history of `lib/models/post.js`, and I found the code was replaced many times:     
https://github.com/hexojs/hexo/commits/master/lib/models/post.js

I think we need to discuss this code.    
(and make enough test cases)

And also `hexo` command *should never* finish normally when a fatal error occurs.

**Comment by yubinbai** [comment]
same stack trace for me

**Comment by NoahDragon** [comment]
Reference to an closed post. #1680

**Comment by Digi59404** [comment]
👍  I also am having this same issue. I updated hexo-renderer-markup to 0.3.0 the latest. This is supposed to have the patched code in #1680. However the issue still exists.

**Comment by NoahDragon** [comment]
@Digi59404 have update Hexo to the latest?

**Comment by Digi59404** [comment]
@NoahDragon Yes, I was at 3.3.5 - I had to downgrade to 3.3.0 in order for it to work.

**Comment by NoahDragon** [comment]
@Digi59404 looks like some fix in between the 3.3.0 and 3.3.5 cause this issue.

**Comment by NoahDragon** [comment]
@seaoak Could you please review the fix https://github.com/hexojs/hexo/pull/2656 ?
